pkg/cache/sql/informer/factory/informer_factory.go

```diff
@@ -149,3 +164,23 @@ func (f *CacheFactory) Reset() error {
 
 	return nil
 }
+
+// checkWatchability checks whether the specified gvk can be watched by this cache factory
+func (f *CacheFactory) checkWatchability(gvk schema.GroupVersionKind) error {
```

Comment:
The performance of this could be a bit rough because this will list all of the resources in a group. In addition, it doesn't cache the response for the given resource type, so if the UI (for example) attempts to watch this resource it will continually attempt to evaluate each of these types. 

I'm wondering if alternatively we could use the `SetWatchErrorHandler` method of the [SharedInformer](https://pkg.go.dev/k8s.io/client-go/tools/cache#SharedInformer) and to check for a [IsMethodNotSupported](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/errors#IsMethodNotSupported) error. This might be a little more complicated, but I think this also could be a bit more accurate since it's parsed from the direct response of the APIserver rather than the openapi schema (which may be inaccurate). WDYT?

Reply:
While I agree that caching would be useful to avoid the API hit, there shouldn't be many `APIResources` in the `apiResourceList`, this is basically getting all the CRDs in the group?

Reply:
If we're going to assume that the OpenAPI specs maybe wrong, we should figure out a plan to fix this.

Reply:
>While I agree that caching would be useful to avoid the API hit, there shouldn't be many APIResources in the apiResourceList, this is basically getting all the CRDs in the group?

That's correct. For a default rancher install the group/version pair with the most resources is `management.cattle.io/v3` with ~50 entries. Don't know how many more resources you could get (eg: with crossplane). Though that will be multiplied by how many cache factory UI is asking (IIUC).

The client-go library has code for caching those calls if that's something we deem important enough: (eg: https://github.com/kubernetes/client-go/blob/master/discovery/discovery_client.go#L105 or https://github.com/kubernetes/client-go/tree/master/discovery/cached). Would have to make sure we invalidate the cache once in a while because the set of api resources can change over time (eg: new CRD, new apiservice, etc).

> rather than the openapi schema (which may be inaccurate)

What do you mean by openapi here? OpenAPI isn't being hit by the code below, just the discovery API.

As far as accuracy goes, it is possible to implement a custom apiserver, that lists `watch` as a supported verb, aggregate it to the main apiserver and have a watch fail. But at this point the contract between main apiserver / custom apiserver is broken so I doubt we could even trust the error response (the buggy apiserver could just send a 500 instead of methodnotsupported :shrug: .

Reply:
> While I agree that caching would be useful to avoid the API hit, there shouldn't be many APIResources in the apiResourceList, this is basically getting all the CRDs in the group?

I think @tomleb  gave a good response to you below, but there can be a couple of resources in the same group, and keep in mind that the call will be made for each of them. 

> I had a deeper look, we're already caching the Cache for the gvk?

Only if the cache was succesfully created: https://github.com/rancher/lasso/blob/6ae667e0fe659adc9f5b77c3adac1e2780555465/pkg/cache/sql/informer/factory/informer_factory.go#L103. My comment here is dealing with the failure case which is not cached (at least as far as I can see).

> If we're going to assume that the OpenAPI specs maybe wrong, we should figure out a plan to fix this.

@tomleb explained this part as well, but an API server can be misconfigured to identify methods it doesn't actually serve. If this is done by a third party (or closed source) application we would have no recourse to fixing this. 

Now it could be argued that my solution depends on us being able to figure out if watch is/is not supported, so there's some room to discuss there, but given that the user can't disable this logic I think that's worth considering.

> OpenAPI isn't being hit by the code below, just the discovery API.

My fault, this would just be the discovery endpoints as @tomleb noted.

> I doubt we could even trust the error response (the buggy apiserver could just send a 500 instead of methodnotsupported 🤷 .)

This is a good point, however, we could parse out more than this specific error (i.e. look for retryable error codes and only retry in those cases, so don't retry with a generic 500). This would still require the APIServer to configure itself correctly, but this would limit our retry action (which is a call to an API) to cases where we had some indication that the issue could be resolved.

Reply:
Valid comments, thank you.

I pushed a completely different approach now so I am resolving this.